### PR TITLE
274 use output index in state trasition verifier instead

### LINKED ISSUE #274: Use `outputIndexOfInputTxs` instead of `inputUtxosPos` for `IStateTransitionVerifier`
### What
Current `IStateTransitionVerifier` interfaces requires the `utxoPos` as input. However, it seems like what it really needs are the `outputIndex` for `inputTx` so the contract can know which output would be used as input.

https://github.com/omisego/plasma-contracts/blob/master/plasma_framework/contracts/src/exits/interfaces/IStateTransitionVerifier.sol#L12

### Why the change
Using `utxoPos` would confuse reader whether a more global plasma status is checked inside or not. It is not designed to be checked there so removal of unnecessary info would make the intention more clear (IMHO).

### PR DESCRIPTION
**Pull request by boolafish**
### Notes
- `utxoPos` --> `outputIndex` in StateTransitionVerifier: closes #274 
- use `StateTransitionVerifierMock` and replace the existing `StateTransitionVerifierAccept/Reject/Reverse` to simplify tests
- add tests in startIFE testing the expected args is called with the stateTransitionVerifier

### Test
```
Contract: PaymentInFlightExitRouter
    startInFlightExit
      when calling start in-flight exit succeed with valid arguments
        ✓ should call the StateTransitionVerifier with correct arguments (745ms)
```

### CODE REVIEW

File: plasma_framework/contracts/mocks/exits/StateTransitionVerifierMock.sol [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    bool public shouldRevert;
+    Args public expectedArgs;
+
+    struct Args {
+        bytes inFlightTx;
+        bytes[] inputTxs;
+        uint16[] outputIndexOfInputTxs;
+    }
+
+    /** mock what would "isCorrectStateTransition()" returns */
```
Comment (pgebal) [inline plasma_framework/contracts/mocks/exits/StateTransitionVerifierMock.sol]: We can remove 'would'.

File: plasma_framework/contracts/mocks/exits/StateTransitionVerifierMock.sol [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    function mockResult(bool result) public {
+        expectedResult = result;
+    }
+
+    /** when called, the "isCorrectStateTransition" function would always revert on purpose */
```
Comment (pgebal) [inline plasma_framework/contracts/mocks/exits/StateTransitionVerifierMock.sol]: would always revert -> reverts

File: plasma_framework/contracts/mocks/exits/StateTransitionVerifierMock.sol [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        // only run the check when "shouldVerifyArgumentEqauals" is called
+        if (expectedArgs.inFlightTx.length > 0) {
+            require(keccak256(inFlightTx) == keccak256(expectedArgs.inFlightTx), "in-flight tx is not as expected");
+
+            require(inputTxs.length == expectedArgs.inputTxs.length, "input txs array length mismatch expected data");
+            for (uint i = 0; i < expectedArgs.inputTxs.length; i++) {
+                require(keccak256(inputTxs[i]) == keccak256(expectedArgs.inputTxs[i]), "input tx is not as expected");
+            }
+
+            require(outputIndexOfInputTxs.length == expectedArgs.outputIndexOfInputTxs.length, "outputIndex array length mismatch expected data");
```
Comment (pgebal) [inline plasma_framework/contracts/mocks/exits/StateTransitionVerifierMock.sol]: mismatch -> mismatches

File: plasma_framework/contracts/src/exits/interfaces/IStateTransitionVerifier.sol [hunk 4]
```diff
@@ -9,7 +9,7 @@ interface IStateTransitionVerifier {
     function isCorrectStateTransition(
         bytes calldata inFlightTx,
         bytes[] calldata inputTxs,
-        uint256[] calldata inputUtxosPos
+        uint16[] calldata outputIndexOfInputTxs
```
Comment (pgebal) [inline plasma_framework/contracts/src/exits/interfaces/IStateTransitionVerifier.sol]: Could you add a doc for this function? `@param`

File: plasma_framework/test/src/exits/payment/PaymentStartInFlightExit.test.js [hunk 5]
```diff
@@ -167,6 +166,22 @@ contract('PaymentInFlightExitRouter', ([_, alice, richFather, carol]) => {
...
             });
 
+            it('should call the StateTransitionVerifier with correct arguments', async () => {
+                const expectedArgs = {
+                    inFlightTx: this.args.inFlightTx,
+                    inputTxs: this.args.inputTxs,
+                    outputIndexOfInputTxs: this.args.inputUtxosPos.map(utxo => new UtxoPos(utxo).outputIndex),
+                };
+
+                // test would fail if called but not with the expected arguments
```
Comment (pgebal) [inline plasma_framework/test/src/exits/payment/PaymentStartInFlightExit.test.js]: if called with unexpected arguments

**Comment by boolafish** [comment]
close this in flavor of #285

File: plasma_framework/contracts/src/exits/interfaces/IStateTransitionVerifier.sol (hunk 4 above)
Comment (boolafish) [reply plasma_framework/contracts/src/exits/interfaces/IStateTransitionVerifier.sol]: okay, I add the doc as following:
I renamed `inFlightTx` to `txBytes` as from this verifier's point of view, it does not care that transaction is in-flight exiting or not.
```
/**
    * @notice Verifies state transition logic.
    * @param txBytes the targeting transaction that is doing the state transition verification
    * @param inputTxs input transactions of the targeting transaction checking state transition
    * @param outputIndexOfInputTxs the output index of the input txs that the transaction input is pointing to
    */
    function isCorrectStateTransition(
        bytes calldata txBytes,
        bytes[] calldata inputTxs,
        uint16[] calldata outputIndexOfInputTxs
    )
        external
        view
        returns (bool);
```

**Pull request merged**